_posts/2021-03-19-forecasting-time-series-with-torch_4/forecasting_4.Rmd

```diff
@@ -0,0 +1,570 @@
+    
+    # concatenate input and context
+    # NOTE: this repeating is done to compensate for the absence of an embedding module
+    # that, in NLP, would give x a higher proportion in the concatenation
+    x_rep <- x$repeat_interleave(dim(context)[3], 3) 
+    rnn_input <- torch_cat(list(x_rep, context), dim = 3)
+    
+    # resulting shapes: (bs, 1, hidden_size) and (1, bs, hidden_size)
+    rnn_out <- self$rnn(rnn_input, state)
+    rnn_output <- rnn_out[[1]]
+    next_hidden <- rnn_out[[2]]
+    
+    mlp_input <- torch_cat(list(rnn_output$squeeze(2), context$squeeze(2), x$squeeze(2)), dim = 2)
+    
+    output <- self$linear(mlp_input)
+    
+    # shapes: (bs, 1) and (1, bs, hidden_size)
+    list(output, next_hidden)
+  }
+  
+)
+
+```
+
+### `seq2seq` module
+
+The `seq2seq` module is basically unchanged (apart from the fact that now, it allows for attention module configuration). For a detailed explanation of what happens here, please consult the [previous post](https://blogs.rstudio.com/ai/posts/2021-03-16-forecasting-time-series-with-torch_3/).
+
+```{r}
+seq2seq_module <- nn_module(
+  
+  initialize = function(type, input_size, hidden_size, attention_type, attention_size, n_forecast, 
+                        num_layers = 1, encoder_dropout = 0) {
+    
+    self$encoder <- encoder_module(type = type, input_size = input_size, hidden_size = hidden_size,
+                                   num_layers, encoder_dropout)
+    self$decoder <- decoder_module(type = type, input_size = 2 * hidden_size, hidden_size = hidden_size,
+                                   attention_type = attention_type, attention_size = attention_size, num_layers)
+    self$n_forecast <- n_forecast
+    
+  },
+  
+  forward = function(x, y, teacher_forcing_ratio) {
+    
+    outputs <- torch_zeros(dim(x)[1], self$n_forecast)$to(device = device)
+    encoded <- self$encoder(x)
+    encoder_outputs <- encoded[[1]]
+    hidden <- encoded[[2]]
+    # list of (batch_size, 1), (1, batch_size, hidden_size)
+    out <- self$decoder(x[ , n_timesteps, , drop = FALSE], hidden, encoder_outputs)
+    # (batch_size, 1)
+    pred <- out[[1]]
+    # (1, batch_size, hidden_size)
+    state <- out[[2]]
+    outputs[ , 1] <- pred$squeeze(2)
+    
+    for (t in 2:self$n_forecast) {
+      
+      teacher_forcing <- runif(1) < teacher_forcing_ratio
+      input <- if (teacher_forcing == TRUE) pred$unsqueeze(3) else y[ , t - 1]
+      out <- self$decoder(pred$unsqueeze(3), state, encoder_outputs)
+      pred <- out[[1]]
+      state <- out[[2]]
+      outputs[ , t] <- pred$squeeze(2)
+      
+    }
+    
+    outputs
+  }
+  
+)
+
+
+```
+
+When instantiating the top-level model, we now have an additional choice: that between additive and multiplicative attention. In the "accuracy" sense of performance, my tests did not show any differences; in the "speed" sense though, the multiplicative variant is a lot faster.
```

Comment:
```suggestion
When instantiating the top-level model, we now have an additional choice: that between additive and multiplicative attention. In the "accuracy" sense of performance, my tests did not show any differences. However, the multiplicative variant is a lot faster.
```

Reply:
oh, I liked that ...? No? OK... with your suggestion, there is still a hint at that ever-recurring ambiguity in ML talk ...